refactor deploy html

bcwu opened the pull request:
<!--- Provide a general summary of your changes in the title. -->

## Description
- Refactor `deploy html` based on the new Manifest and Bundle classes
- Add missing unit test coverage for `deploy html`

bcwu commented:
fixes https://github.com/rstudio/rsconnect-python/issues/361

bcwu commented:
also, fixes https://github.com/rstudio/rsconnect-python/issues/359

kgartland-rstudio commented:
Can deploy and render...

single html file: deploy by directory, no entrypoint
- [x] index.html
- [x] blue.html (manifest contains proper entrypoint)
- [x] no html (caught Error: No valid entrypoint found.)

single html file deploy by directory with entrypoint
- [x] index.html
- [x] blue.html
- [x] no html (connect provides handy download page)
![Screen Shot 2023-03-09 at 3 28 15 PM](https://user-images.githubusercontent.com/41300090/224148211-37a21e35-a581-42e5-95cf-e8407841a9ac.png)
- ❌  wrong entrypoint


single html file: deploy by filename
- [x] index.html
- [x] blue.html
- [x] no html
- [x] wrong entrypoint

single html file: deploy by filename with entrypoint
- [x] index.html
- [x] blue.html
- [x] no html
- ❌  wrong entrypoint

multi html deploy: deploy by directory, no entrypoint
- [x] index.html
- [x] blue.html (manifest contains proper entrypoint)
- [x] no html (caught Error: No valid entrypoint found.)

multi html deploy: deploy by file
- ❌ specifying two files

kgartland-rstudio commented:
When specifying an entrypoint that doesn't exist we don't catch the error and give a reasonable response, instead we throw a NoneType Err:

In this example the `--entrypoint blue` doesn't exist.
```
> rsconnect deploy html html-multi-file/index.html --entrypoint blue -n dogfood -t "multiple html" -N
Validating server... 	[OK]
Validating app mode... 	[OK]
Making bundle ... 	[ERROR]: expected str, bytes or os.PathLike object, not NoneType
Traceback (most recent call last):
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/main.py", line 95, in wrapper
    result = func(*args, **kwargs)
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/main.py", line 1188, in deploy_html
    ce.validate_server()
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/log.py", line 186, in wrapper
    result = method(self, *args, **kw)
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/api.py", line 636, in make_bundle
    bundle = func(*args, **kwargs)
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/bundle.py", line 925, in make_html_bundle
    manifest = create_html_manifest(**locals())
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/bundle.py", line 890, in create_html_manifest
    excludes.extend(list_environment_dirs(deploy_dir))
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/bundle.py", line 792, in list_environment_dirs
    path = join(directory, name)
  File "/Users/kgartland/.pyenv/versions/3.8.2/lib/python3.8/posixpath.py", line 76, in join
    a = os.fspath(a)
TypeError: expected str, bytes or os.PathLike object, not NoneType
Internal error: expected str, bytes or os.PathLike object, not NoneType
```

kgartland-rstudio commented:
When specifying two files in an html deployment, from outside the directory, we state:
No such file or directory, even when the file exists:

With this file structure:
```
-> tree html-multi-file/
html-multi-file/
├── index.html
├── package.json
├── rsconnect-python
│   └── html-multi-file.json
├── sandbox.config.json
└── second.html
```

From outside the directory:
```
> rsconnect deploy html html-multi-file/index.html html-multi-file/second.html -n dogfood -t "multiple html" -N
Validating server... 	[OK]
Validating app mode... 	[OK]
Making bundle ... 	[ERROR]: Unable to include the file second.html in the bundle: No such file or directory
Error: Unable to include the file second.html in the bundle: No such file or directory
```

When specifying two files in an html deployment, from inside the directory, we state: 
Internal error: 'second.html' does not start with [PATH]:

```
> rsconnect deploy html index.html second.html -n dogfood -t "multiple html" -N
Validating server... 	[OK]
Validating app mode... 	[OK]
Making bundle ... 	[ERROR]: 'second.html' does not start with '/Users/kgartland/work/connect/test/rsconnect-python/bundles/html-multi-file'
Traceback (most recent call last):
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/main.py", line 95, in wrapper
    result = func(*args, **kwargs)
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/main.py", line 1188, in deploy_html
    ce.validate_server()
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/log.py", line 186, in wrapper
    result = method(self, *args, **kw)
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/api.py", line 636, in make_bundle
    bundle = func(*args, **kwargs)
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/bundle.py", line 941, in make_html_bundle
    return bundle.to_file()
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/bundle.py", line 274, in to_file
    rel_path = Path(fp).relative_to(self.deploy_dir) if flatten_to_deploy_dir else None
  File "/Users/kgartland/.pyenv/versions/3.8.2/lib/python3.8/pathlib.py", line 899, in relative_to
    raise ValueError("{!r} does not start with {!r}"
ValueError: 'second.html' does not start with '/Users/kgartland/work/connect/test/rsconnect-python/bundles/html-multi-file'
Internal error: 'second.html' does not start with '/Users/kgartland/work/connect/test/rsconnect-python/bundles/html-multi-file'
```

bcwu commented:
> When specifying an entrypoint that doesn't exist we don't catch the error and give a reasonable response, instead we throw a NoneType Err:

fixed by [f5a2146](https://github.com/rstudio/rsconnect-python/pull/362/commits/f5a214624abaf519d30fab1580edf2386116e6bf)



> When specifying two files in an html deployment, from outside the directory, we state: No such file or directory, even when the file exists:
> 
> With this file structure:
> 
> ```
> -> tree html-multi-file/
> html-multi-file/
> ├── index.html
> ├── package.json
> ├── rsconnect-python
> │   └── html-multi-file.json
> ├── sandbox.config.json
> └── second.html
> ```
> 
fixed by [2f8703a](https://github.com/rstudio/rsconnect-python/pull/362/commits/2f8703a597aeccbed1512cd610ed0353e420410b)

kgartland-rstudio commented:
> > When specifying an entrypoint that doesn't exist we don't catch the error and give a reasonable response, instead we throw a NoneType Err:
> 
> fixed by [f5a2146](https://github.com/rstudio/rsconnect-python/pull/362/commits/f5a214624abaf519d30fab1580edf2386116e6bf)
> 
> > When specifying two files in an html deployment, from outside the directory, we state: No such file or directory, even when the file exists:
> > With this file structure:
> > ```
> > -> tree html-multi-file/
> > html-multi-file/
> > ├── index.html
> > ├── package.json
> > ├── rsconnect-python
> > │   └── html-multi-file.json
> > ├── sandbox.config.json
> > └── second.html
> > ```
> 
> fixed by [2f8703a](https://github.com/rstudio/rsconnect-python/pull/362/commits/2f8703a597aeccbed1512cd610ed0353e420410b)

These fixes look good.

kgartland-rstudio commented:
Windows testing.  When multiple files are in the target directory it fails with this error:

```
> rsconnect deploy html .\html\html-single-file\ -s https://rsc.radixu.com/ -k bKnYIJiJasSulZcbUlhZ4UkMYdL7YCWq
Validating server...    [OK]
Validating app mode...  [OK]
Making bundle ...       [ERROR]: 'html\\html-single-file\\nothin.html' is not in the subpath of 'C:\\Users\\User\\html\\html-single-file' OR one path is relative and the other is absolute.
Traceback (most recent call last):
  File "C:\Users\User\AppData\Local\Programs\Python\Python310\lib\site-packages\rsconnect\main.py", line 95, in wrapper
    result = func(*args, **kwargs)
  File "C:\Users\User\AppData\Local\Programs\Python\Python310\lib\site-packages\rsconnect\main.py", line 1190, in deploy_html
    .make_bundle(
  File "C:\Users\User\AppData\Local\Programs\Python\Python310\lib\site-packages\rsconnect\log.py", line 186, in wrapper
    result = method(self, *args, **kw)
  File "C:\Users\User\AppData\Local\Programs\Python\Python310\lib\site-packages\rsconnect\api.py", line 636, in make_bundle
    bundle = func(*args, **kwargs)
  File "C:\Users\User\AppData\Local\Programs\Python\Python310\lib\site-packages\rsconnect\bundle.py", line 940, in make_html_bundle
    return bundle.to_file()
  File "C:\Users\User\AppData\Local\Programs\Python\Python310\lib\site-packages\rsconnect\bundle.py", line 274, in to_file
    rel_path = Path(fp).relative_to(self.deploy_dir) if flatten_to_deploy_dir else None
  File "C:\Users\User\AppData\Local\Programs\Python\Python310\lib\pathlib.py", line 818, in relative_to
    raise ValueError("{!r} is not in the subpath of {!r}"
ValueError: 'html\\html-single-file\\nothin.html' is not in the subpath of 'C:\\Users\\User\\html\\html-single-file' OR one path is relative and the other is absolute.
Internal error: 'html\\html-single-file\\nothin.html' is not in the subpath of 'C:\\Users\\User\\html\\html-single-file' OR one path is relative and the other is absolute.
```

Directory structure:

```
    Directory: C:\Users\User\html\html-single-file


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d-----         3/10/2023   6:58 AM                rsconnect-python
-a----         3/10/2023   6:58 AM            144 index.html
------         3/10/2023   6:58 AM              0 nothin.html
```

kgartland-rstudio commented:
When specifying an entrypoint where multiple html files exist...

1. do we expect to require the path to the file? or just the file name itself?  Since we're giving the path to the deployment directory, I would expect that the path wouldn't be required in the entrypoint arg.

```
> rsconnect deploy html bundles/html-multi-file/ --entrypoint second.html -n dogfood
Validating server... 	[OK]
Validating app mode... 	[OK]
Making bundle ... 	[ERROR]: Entrypoint second.html does not exist.
Error: Entrypoint second.html does not exist.
```
Directory structure:
```
> tree bundles/html-multi-file/
bundles/html-multi-file/
├── index.html
├── package.json
├── rsconnect-python
│   └── html-multi-file.json
├── sandbox.config.json
└── second.html
```

2. When you _do_ specify the full path in the entrypoint arg, it fails:
```
> rsconnect deploy html bundles/html-multi-file/ --entrypoint bundles/html-multi-file/second.html -n dogfood
Validating server... 	[OK]
Validating app mode... 	[OK]
Making bundle ... 	[ERROR]: 'bundles/html-multi-file/index.html' does not start with '/Users/kgartland/work/connect/test/rsconnect-python/bundles/html-multi-file'
Traceback (most recent call last):
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/main.py", line 95, in wrapper
    result = func(*args, **kwargs)
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/main.py", line 1188, in deploy_html
    ce.validate_server()
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/log.py", line 186, in wrapper
    result = method(self, *args, **kw)
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/api.py", line 636, in make_bundle
    bundle = func(*args, **kwargs)
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/bundle.py", line 940, in make_html_bundle
    return bundle.to_file()
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/bundle.py", line 274, in to_file
    rel_path = Path(fp).relative_to(self.deploy_dir) if flatten_to_deploy_dir else None
  File "/Users/kgartland/.pyenv/versions/3.8.2/lib/python3.8/pathlib.py", line 899, in relative_to
    raise ValueError("{!r} does not start with {!r}"
ValueError: 'bundles/html-multi-file/index.html' does not start with '/Users/kgartland/work/connect/test/rsconnect-python/bundles/html-multi-file'
Internal error: 'bundles/html-multi-file/index.html' does not start with '/Users/kgartland/work/connect/test/rsconnect-python/bundles/html-multi-file'
```

kgartland-rstudio commented:
It looks like this part is still broken:

> do we expect to require the path to the file? or just the file name itself? Since we're giving the path to the deployment directory, I would expect that the path wouldn't be required in the entrypoint arg.
```
> rsconnect deploy html bundles/html-multi-file/ --entrypoint second.html -n dogfood
Validating server... 	[OK]
Validating app mode... 	[OK]
Making bundle ... 	[ERROR]: Entrypoint second.html does not exist.
Error: Entrypoint second.html does not exist.
```

bcwu commented:
> It looks like this part is still broken:
> 
> > do we expect to require the path to the file? or just the file name itself? Since we're giving the path to the deployment directory, I would expect that the path wouldn't be required in the entrypoint arg.
> 
> ```
> > rsconnect deploy html bundles/html-multi-file/ --entrypoint second.html -n dogfood
> Validating server... 	[OK]
> Validating app mode... 	[OK]
> Making bundle ... 	[ERROR]: Entrypoint second.html does not exist.
> Error: Entrypoint second.html does not exist.

The full path has to be specified for the entrypoint if you are outside the folder. So in this case:

```
rsconnect deploy html bundles/html-multi-file/ --entrypoint bundles/html-multi-file/second.html -n dogfood
```

The other way is to do relative deploy, like so:

```
cd html bundles/html-multi-file/
rsconnect deploy html ./ --entrypoint second.html -n dogfood
```

bcwu commented:
> We'll at least have to document around that.

Added documentation on the deploy html behavior in the [changelog](https://github.com/rstudio/rsconnect-python/blob/bcwu-refactor-deploy-html/CHANGELOG.md) 

Let me know if there's anything that's missed.

kgartland-rstudio commented on CHANGELOG.md:
```diff
@@ -4,6 +4,68 @@ All notable changes to this project will be documented in this file.
+- `deploy html` was refactored. Its behavior is described as 
+follows:
+
+### deploy html
```
this feels like something that needs to be in our [docs](https://docs.posit.co/rsconnect-python/#static-snapshot-deployment), not just in the changelog.

bcwu replied:
Will open a separate PR for Connect docs. Will leave the changes in the changelog for now, unless there's anything we should cut and move over to the docs.

kgartland-rstudio commented:
Testing on Windows is complete.  I think we're good to merge 👍

kgartland-rstudio commented on rsconnect/bundle.py:
```diff
@@ -817,61 +840,111 @@ def make_api_manifest(
+    image: str = None,
+) -> typing.IO[bytes]:
+    """
+    Create an voila bundle, given a path and/or entrypoint.
```
one last silly nitpick. this probably needs to be updated to html.
